Algorithmic improvements to AL mortar mechanical contact

recuero opened the pull request:
Closes #25404

recuero committed:
Use LMs from previous AL iteration.

For problems with large nonlinearities, not jump starting the contact
problem with previous step's contact tractions means that the contact
state needs to be solved for all over again.

recuero committed:
Adaptivity updates.

Add Bussetta adaptivity for normal contact.

Update 'MOOSE' adaptivity for normal and frictional contact. Now, the local
tolerance criterion is used to stop adapting.

It remains: (1) Add Busetta for friction (2) Test Bussetta for normal and
frictional contact. We should provide the two ways of adapting those coefficients.

recuero committed:
Cleanup adaptivity.

Provide two ways for the user to adapt the normal/frictional penalty.
Provide a real that sets a maximum penalty factor based on the initial
penalty value. This is handy to avoid too-large penalty factors that
may compromise the solution of the system.

recuero commented:
Address https://github.com/idaholab/moose/pull/24349#pullrequestreview-1592516437

recuero committed:
Address Ben's comments from #24349

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1750705/) on 4c9fd3f : invalidated by @recuero

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1750697/) on 4c9fd3f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25405/site)

This comment will be updated on new commits.

recuero committed:
Update tests to new default settings.

Add AL three-dimensional test with friction.

dschwen commented on modules/contact/test/tests/3d-mortar-contact/tests:
```diff
@@ -183,4 +183,18 @@
+    type = 'Exodiff'
+    input = 'frictional-mortar-3d-al.i'
+    exodiff = 'frictional-mortar-3d-al_out.e'
+    requirement = 'The system shall solve a 3D frictional problem with mortar constraint using a '
```
```suggestion
    requirement = 'The system shall be able to solve a 3D frictional problem with mortar constraint using a '
```

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1752390/) on d8ec597 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>a1e20d</th>
      <th colspan=4>#25405 d8ec59</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.86%</td>
      <td>89.18%</td>
      <td>-0.69%</td>
      <td>64.15%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3830</td>
      <td>3873</td>
      <td>+43</td>
      <td>68</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>432</td>
      <td>470</td>
      <td>+38</td>
      <td>38</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25405/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25405/coverage/contact)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25405/coverage/xfem)

</details>

## Warnings

- `contact` new line coverage rate 64.15% is less than the suggested 90.0%



This comment will be updated on new commits.

recuero commented:
Need to add coverage.

dschwen reviewed (commented):
Looks fine to me, just some minor comments.

recuero commented on modules/contact/include/userobjects/PenaltyFrictionUserObject.h:
```diff
@@ -63,6 +63,9 @@ class PenaltyFrictionUserObject : virtual public PenaltyWeightedGapUserObject,
   const Real _friction_coefficient;
 
+  /// Map from degree of freedom to current and old step slip
+  std::unordered_map<const DofObject *, std::pair<TwoVector, TwoVector>> _dof_to_step_slip;
```
Agreed. We may see other bottlenecks in the profiler report too.

recuero committed:
Fix action -> Increase converage

recuero committed:
Address Daniel's review comments

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1753643/) on 20de956 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25405/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25405/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a1e20d32d7c0cf0e6d911a534c5ec640fc7306cd`

recuero committed:
Attempt at covering more cases for adaptive penalty

recuero committed:
Save step slip in a good place.

This misbehavior was revealed as a "parallel" mismatch in the frictional
traction. It gives consistent parallel results now.

Reduce runtime of Bussetta's test. It was timing out in debug mode.

dschwen commented:
Merge commit?

recuero commented:
Topher's residual tags PR created a conflict.

recuero commented:
I'll rebase locally. This on-PR merge commit is a mess.

recuero committed:
Merge commit and remaining files

recuero commented:
@dschwen this should be ready for another review or approval.
